include/envoy/registry/registry.h

```diff
@@ -18,6 +18,51 @@
 namespace Envoy {
 namespace Registry {
 
+template <class Base> class FactoryRegistry;
+
+/**
+ * Helper class to call `registeredNames` for a specialized
+ * FactoryRegistry.
+ */
+class BaseFactoryCategoryNames {
+public:
+  virtual ~BaseFactoryCategoryNames() = default;
+  virtual std::vector<absl::string_view> registeredNames() const PURE;
+};
+
+template <class Base> class FactoryCategoryNames : public BaseFactoryCategoryNames {
+public:
+  using FactoryRegistry = Envoy::Registry::FactoryRegistry<Base>;
+
+  std::vector<absl::string_view> registeredNames() const override {
+    return FactoryRegistry::registeredNames();
+  }
+};
+
+/**
+ * FactoryCategoryRegistry registers factory registries by their
+ * declared category. The category is exposed by a static category()
+ * method on the factory base type.
+ */
+class FactoryCategoryRegistry {
+public:
+  using MapType = absl::flat_hash_map<std::string, BaseFactoryCategoryNames*>;
+
+  static MapType& factories() {
+    static auto* factories = new MapType();
+    return *factories;
+  }
+
+  static bool isRegistered(absl::string_view category) { return factories().contains(category); }
+
+  static void registerCategory(absl::string_view category, BaseFactoryCategoryNames* factoryNames) {
+    auto result = factories().emplace(std::make_pair(category, factoryNames));
+    if (!result.second) {
+      throw EnvoyException(fmt::format("Double registration for category: '{}'", category));
```

Comment:
Please add a test case for this somewhere, though can this actually happen since you already check if it's registered? Can this just be an assert?

Reply:
I can make it an assert (using the `RELEASE_ASSERT` macro?). I'll look at the `FactoryRegistry` tests to see whether I can knock up a test for this case.

Reply:
I restructured a bit so only `RegisterFactory` can register a category. So this is probably impossible to get wrong now?